backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/serialization/IrModuleDeserializer.kt

```diff
@@ -0,0 +1,1130 @@
+            -> deserializeEnumConstructorCall(proto.enumConstructorCall, start, end, type)
+            FUNCTION_REFERENCE
+            -> deserializeFunctionReference(proto.functionReference, start, end, type)
+            GET_ENUM_VALUE
+            -> deserializeGetEnumValue(proto.getEnumValue, start, end, type)
+            GET_CLASS
+            -> deserializeGetClass(proto.getClass, start, end, type)
+            GET_FIELD
+            -> deserializeGetField(proto.getField, start, end, type)
+            GET_OBJECT
+            -> deserializeGetObject(proto.getObject, start, end, type)
+            GET_VALUE
+            -> deserializeGetValue(proto.getValue, start, end, type)
+            INSTANCE_INITIALIZER_CALL
+            -> deserializeInstanceInitializerCall(proto.instanceInitializerCall, start, end)
+            PROPERTY_REFERENCE
+            -> deserializePropertyReference(proto.propertyReference, start, end, type)
+            RETURN
+            -> deserializeReturn(proto.`return`, start, end, type)
+            SET_FIELD
+            -> deserializeSetField(proto.setField, start, end)
+            SET_VARIABLE
+            -> deserializeSetVariable(proto.setVariable, start, end)
+            STRING_CONCAT
+            -> deserializeStringConcat(proto.stringConcat, start, end, type)
+            THROW
+            -> deserializeThrow(proto.`throw`, start, end, type)
+            TRY
+            -> deserializeTry(proto.`try`, start, end, type)
+            TYPE_OP
+            -> deserializeTypeOp(proto.typeOp, start, end, type)
+            VARARG
+            -> deserializeVararg(proto.vararg, start, end, type)
+            WHEN
+            -> deserializeWhen(proto.`when`, start, end, type)
+            WHILE
+            -> deserializeWhile(proto.`while`, start, end, type)
+            OPERATION_NOT_SET
+            -> error("Expression deserialization not implemented: ${proto.operationCase}")
+        }
+
+    private fun deserializeExpression(proto: KonanIr.IrExpression): IrExpression {
+        val start = proto.coordinates.startOffset
+        val end = proto.coordinates.endOffset
+        val type = deserializeIrType(proto.type)
+        val operation = proto.operation
+        val expression = deserializeOperation(operation, start, end, type)
+
+        logger.log { "### Deserialized expression: ${ir2string(expression)} ir_type=$type" }
+        return expression
+    }
+
+    private fun deserializeIrTypeParameter(
+        proto: KonanIr.IrTypeParameter,
+        start: Int,
+        end: Int,
+        origin: IrDeclarationOrigin
+    ): IrTypeParameter {
+        val symbol = deserializeIrSymbol(proto.symbol) as IrTypeParameterSymbol
+        val name = Name.identifier(proto.name)
+        val variance = deserializeIrTypeVariance(proto.variance)
+
+        val parameter = symbolTable.declareGlobalTypeParameter(UNDEFINED_OFFSET, UNDEFINED_OFFSET, irrelevantOrigin,
+            symbol.descriptor, { symbol ->
```

Comment:
Creating the symbol twice (here and in `deserializeIrSymbol` above) looks somewhat dirty. It is worth at least adding a TODO.